### PR TITLE
Fix Fk verification and update queries to accommodate for bindVariables being NULL

### LINKED ISSUE #14075: Bug Report: Setting fk column to NULL in prepared statements
### Overview of the Issue

This bug was found using the fuzzer introduced in #13980.

The test with prepared statements started failing when NULL values were enabled in it. It was noticed that running updates such that the foreign key related column was set to NULL causes Vitess to not handle it correctly. - 
```go
=== RUN   TestFkFuzzTest/Unsharded_-_Single_Thread_-_Balanced_Inserts_and_Updates_Prepared_-_true
    mysql.go:232: Vitess and MySQL are not erroring the same way.
        Vitess error: <nil>
        MySQL error: Cannot delete or update a parent row: a foreign key constraint fails (`ks`.`fk_t13`, CONSTRAINT `fk_t13_ibfk_1` FOREIGN KEY (`col`) REFERENCES `fk_t11` (`col`) ON DELETE RESTRICT ON UPDATE RESTRICT) (errno 1451) (sqlstate 23000) during query: execute stmt_update using @col, @id
E0922 11:15:13.251375   71283 fk_fuzz_test.go:503] Failing query - [prepare stmt_update from 'update fk_t10 set col = ? where id = ?' SET @id = 3 SET @col = NULL execute stmt_update using @col, @id]
```

It seems that MySQL rejects the query, but Vitess doesn't. 

Schema for the tables in question is as follows - 
```go
/*
 *                fk_t10
 *                   │
 * On Delete Cascade │
 * On Update Cascade │
 *                   │
 *                   ▼
 *                fk_t11──────────────────┐
 *                   │                    │
 *                   │                    │ On Delete Restrict
 * On Delete Cascade │                    │ On Update Restrict
 * On Update Cascade │                    │
 *                   │                    │
 *                   ▼                    ▼
 *                fk_t12               fk_t13
 */
```

The data prior to running the query in the database was - 
```
E0922 11:14:29.281132   71283 fk_fuzz_test.go:202] Initial data for fk_t10 -
[[INT64(1) VARCHAR("5")] [INT64(2) VARCHAR("3")] [INT64(3) VARCHAR("2")] [INT64(4) VARCHAR("1")] [INT64(5) VARCHAR("4")] [INT64(6) VARCHAR("2")] [INT64(7) VARCHAR("1")] [INT64(8) NULL] [INT64(9) VARCHAR("2")] [INT64(10) VARCHAR("3")]]
E0922 11:14:29.281149   71283 fk_fuzz_test.go:202] Initial data for fk_t11 -
[[INT64(1) VARCHAR("2")] [INT64(2) VARCHAR("3")] [INT64(5) VARCHAR("1")] [INT64(6) VARCHAR("2")] [INT64(7) VARCHAR("3")] [INT64(8) VARCHAR("1")] [INT64(9) VARCHAR("5")] [INT64(10) VARCHAR("1")]]
E0922 11:14:29.281155   71283 fk_fuzz_test.go:202] Initial data for fk_t12 -
[[INT64(1) NULL] [INT64(3) NULL] [INT64(4) VARCHAR("1")] [INT64(5) NULL] [INT64(6) VARCHAR("3")] [INT64(7) VARCHAR("1")] [INT64(8) VARCHAR("1")] [INT64(9) NULL]]
E0922 11:14:29.281165   71283 fk_fuzz_test.go:202] Initial data for fk_t13 -
[[INT64(3) VARCHAR("2")] [INT64(4) VARCHAR("2")] [INT64(5) VARCHAR("5")] [INT64(6) NULL] [INT64(8) VARCHAR("1")] [INT64(9) VARCHAR("3")]]
```

After running the query, the data in Vitess and MySQL doesn't match - 
```
mysql.go:223: Query (SELECT * FROM fk_t10 ORDER BY id) results mismatched.
        Vitess Results:
        [INT64(1) VARCHAR("5")]
        [INT64(2) VARCHAR("3")]
        [INT64(3) NULL]
        [INT64(4) VARCHAR("1")]
        [INT64(5) VARCHAR("4")]
        [INT64(6) VARCHAR("2")]
        [INT64(7) VARCHAR("1")]
        [INT64(8) NULL]
        [INT64(9) VARCHAR("2")]
        [INT64(10) VARCHAR("3")]
        Vitess RowsAffected: 0
        MySQL Results:
        [INT64(1) VARCHAR("5")]
        [INT64(2) VARCHAR("3")]
        [INT64(3) VARCHAR("2")]
        [INT64(4) VARCHAR("1")]
        [INT64(5) VARCHAR("4")]
        [INT64(6) VARCHAR("2")]
        [INT64(7) VARCHAR("1")]
        [INT64(8) NULL]
        [INT64(9) VARCHAR("2")]
        [INT64(10) VARCHAR("3")]
        MySQL RowsAffected: 0

mysql.go:223: Query (SELECT * FROM fk_t11 ORDER BY id) results mismatched.
        Vitess Results:
        [INT64(1) NULL]
        [INT64(2) VARCHAR("3")]
        [INT64(5) VARCHAR("1")]
        [INT64(6) NULL]
        [INT64(7) VARCHAR("3")]
        [INT64(8) VARCHAR("1")]
        [INT64(9) VARCHAR("5")]
        [INT64(10) VARCHAR("1")]
        Vitess RowsAffected: 0
        MySQL Results:
        [INT64(1) VARCHAR("2")]
        [INT64(2) VARCHAR("3")]
        [INT64(5) VARCHAR("1")]
        [INT64(6) VARCHAR("2")]
        [INT64(7) VARCHAR("3")]
        [INT64(8) VARCHAR("1")]
        [INT64(9) VARCHAR("5")]
        [INT64(10) VARCHAR("1")]
        MySQL RowsAffected: 0
```

### Reproduction Steps

1. Create an unsharded keyspace with foreign keys as stated and Vitess managing the foreign keys.
2. Insert the data into the tables such that the initial state is reached
3. Run the prepared statement queries as stated.
4. See that Vitess doesn't reject the query even though it should due to a contraint failure on t13. 

### Binary Version

```sh
main
```

### Operating System and Environment details

```sh
main
```

### Log Fragments

_No response_

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

This PR fixes the bug as reported in https://github.com/vitessio/vitess/issues/14075.

The code that was written previously assumed that a bindVariable would always be a not NULL value because Vitess doesn't parameterize NULL values to bindVariables. This assumption, however, breaks down when we come to prepared statements. The `?` values are changed to bindVariables which can be set to NULL on execution.

This causes Vitess to create incorrect queries when it is verifying that the child table has any rows for the RESTRICT case, and when cascading SET NULLs. 

Specifically for the case described in #14075, `update fk_t10 set col = :v2 where id = :v1`, before the changes in this PR, when we cascade the changes into `fk_t11` we have to do so with `FOREIGN_KEY_CHECKS=OFF`, and verify the validity of the RESTRICY foreign key on vtgate. To that end, we were producing the query like so - 
```sql
select 1 from fk_t11, fk_t13 
where (fk_t11.col) in ::fkc_vals and (fk_t11.col) not in ((:v2)) and fk_t11.col = fk_t13.col 
limit 1 lock in share mode
```

In this case if `:v2` is NULL on runtime, then the condition `fk_t11.col NOT IN (NULL)` becomes false always, since no value evaluates to true when checked for non-equality with NULL. This means that Vitess would always assume that it is safe to proceed with the update.

This is evidently incorrect. This PR fixes the query we use for this purpose by changing the where condition to also first make sure that the bindVariable is not null. The new query now looks like -
```sql
select 1 from fk_t11, fk_t13 
where (fk_t11.col) in ::fkc_vals and 
(:v2 is null or (fk_t11.col) not in ((:v2))) 
and fk_t11.col = fk_t13.col 
limit 1 lock in share mode
```

Now, the query is correct and it finds the rows that should be blocking the updates. The full plan for the original query after the changes looks like - 
```json
{
          "QueryType": "UPDATE",
          "Original": "update fk_t10 set col = :v2 where id = :v1",
          "Instructions": {
            "OperatorType": "FkCascade",
            "Inputs": [
              {
                "InputName": "Selection",
                "OperatorType": "Route",
                "Variant": "Unsharded",
                "Keyspace": {
                  "Name": "unsharded_fk_allow",
                  "Sharded": false
                },
                "FieldQuery": "select col from fk_t10 where 1 != 1",
                "Query": "select col from fk_t10 where id = :v1 for update",
                "Table": "fk_t10"
              },
              {
                "InputName": "CascadeChild-1",
                "OperatorType": "FKVerify",
                "BvName": "fkc_vals",
                "Cols": [
                  0
                ],
                "Inputs": [
                  {
                    "InputName": "VerifyChild-1",
                    "OperatorType": "Route",
                    "Variant": "Unsharded",
                    "Keyspace": {
                      "Name": "unsharded_fk_allow",
                      "Sharded": false
                    },
                    "FieldQuery": "select 1 from fk_t11, fk_t13 where 1 != 1",
                    "Query": "select 1 from fk_t11, fk_t13 where (fk_t11.col) in ::fkc_vals and (:v2 is null or (fk_t11.col) not in ((:v2))) and fk_t11.col = fk_t13.col limit 1 lock in share mode",
                    "Table": "fk_t11, fk_t13"
                  },
                  {
                    "InputName": "PostVerify",
                    "OperatorType": "FkCascade",
                    "Inputs": [
                      {
                        "InputName": "Selection",
                        "OperatorType": "Route",
                        "Variant": "Unsharded",
                        "Keyspace": {
                          "Name": "unsharded_fk_allow",
                          "Sharded": false
                        },
                        "FieldQuery": "select col from fk_t11 where 1 != 1",
                        "Query": "select col from fk_t11 where (fk_t11.col) in ::fkc_vals for update",
                        "Table": "fk_t11"
                      },
                      {
                        "InputName": "CascadeChild-1",
                        "OperatorType": "Update",
                        "Variant": "Unsharded",
                        "Keyspace": {
                          "Name": "unsharded_fk_allow",
                          "Sharded": false
                        },
                        "TargetTabletType": "PRIMARY",
                        "BvName": "fkc_vals1",
                        "Cols": [
                          0
                        ],
                        "Query": "update /*+ SET_VAR(foreign_key_checks=OFF) */ fk_t12 set col = :v2 where (col) in ::fkc_vals1",
                        "Table": "fk_t12"
                      },
                      {
                        "InputName": "Parent",
                        "OperatorType": "Update",
                        "Variant": "Unsharded",
                        "Keyspace": {
                          "Name": "unsharded_fk_allow",
                          "Sharded": false
                        },
                        "TargetTabletType": "PRIMARY",
                        "Query": "update /*+ SET_VAR(foreign_key_checks=OFF) */ fk_t11 set col = :v2 where (fk_t11.col) in ::fkc_vals",
                        "Table": "fk_t11"
                      }
                    ]
                  }
                ]
              },
              {
                "InputName": "Parent",
                "OperatorType": "Update",
                "Variant": "Unsharded",
                "Keyspace": {
                  "Name": "unsharded_fk_allow",
                  "Sharded": false
                },
                "TargetTabletType": "PRIMARY",
                "Query": "update fk_t10 set col = :v2 where id = :v1",
                "Table": "fk_t10"
              }
            ]
          },
          "TablesUsed": [
            "unsharded_fk_allow.fk_t10",
            "unsharded_fk_allow.fk_t11",
            "unsharded_fk_allow.fk_t12",
            "unsharded_fk_allow.fk_t13"
          ]
        }
```

A similar change to the where condition has also been made to the case where we SET NULLs on the child in case of `ON UPDATE SET NULL`

Post these changes, I have verified the fuzzer (#13980) is happy and doesn't complain anymore! 🚀 
## Related Issue(s)

- Fixes #14075 

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/testdata/foreignkey_cases.json [hunk 1]
```diff
@@ -1364,7 +1453,7 @@
                   "Sharded": false
                 },
                 "TargetTabletType": "PRIMARY",
-                "Query": "update u_multicol_tbl2 set cola = null, colb = null where (cola, colb) in ::fkc_vals and (cola, colb) not in ((1, 2))",
+                "Query": "update u_multicol_tbl2 set cola = null, colb = null where (cola, colb) in ::fkc_vals and (1 is null or (cola, colb) not in ((1, 2)))",
```
Comment (GuptaManan100) [inline go/vt/vtgate/planbuilder/testdata/foreignkey_cases.json:1456]: This change will go away when https://github.com/vitessio/vitess/pull/14059 is merged.

**Review by shlomi-noach — APPROVED** [review]

File: go/vt/vtgate/planbuilder/testdata/foreignkey_cases.json (hunk 1 above)
Comment (GuptaManan100) [reply go/vt/vtgate/planbuilder/testdata/foreignkey_cases.json]: The PR has been merged and the expectations changed back.

File: go/vt/vtgate/planbuilder/operators/update.go [hunk 2]
```diff
@@ -618,7 +638,17 @@ func createFkVerifyOpForChildFKForUpdate(ctx *plancontext.PlanningContext, updSt
...
+		for _, value := range updateValues {
+			finalExpr = &sqlparser.OrExpr{
+				Left: &sqlparser.IsExpr{
+					Left:  value,
+					Right: sqlparser.IsNullOp,
+				},
+				Right: finalExpr,
+			}
+		}
+		whereCond = sqlparser.AndExpressions(whereCond, finalExpr)
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/update.go:651]: nit: looks like similar changes, can this be extracted to a common method?

File: go/vt/vtgate/planbuilder/testdata/foreignkey_cases.json [hunk 3]
```diff
@@ -1390,6 +1479,100 @@
...
+              {
+                "InputName": "Selection",
+                "OperatorType": "Route",
+                "Variant": "Unsharded",
+                "Keyspace": {
+                  "Name": "unsharded_fk_allow",
+                  "Sharded": false
+                },
+                "FieldQuery": "select cola, colb from u_multicol_tbl2 where 1 != 1",
+                "Query": "select cola, colb from u_multicol_tbl2 where (cola, colb) in ::fkc_vals and (:v2 is null or (:v1 is null or (cola, colb) not in ((:v1, :v2)))) for update",
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/testdata/foreignkey_cases.json:1521]: can be simplified to
```
:v2 is null or :v1 is null or (cola, colb) not in ((:v1, :v2))
```

**Comment by harshit-gangal** [comment]
All changes looks good.
do we have multi-column fk in fuzzer test?

**Review by harshit-gangal — APPROVED** [review]

File: go/vt/vtgate/planbuilder/operators/update.go (hunk 2 above)
Comment (GuptaManan100) [reply go/vt/vtgate/planbuilder/operators/update.go]: Fixed 👍

File: go/vt/vtgate/planbuilder/testdata/foreignkey_cases.json (hunk 3 above)
Comment (GuptaManan100) [reply go/vt/vtgate/planbuilder/testdata/foreignkey_cases.json]: The parentheses are added by the AST when it is printed, we can't get rid of them.

**Comment by GuptaManan100** [comment]
Yes, we do. The fuzzer has multi-col foreign key tests as well.

**Pull request merged**